sympy/polys/matrices/sdm.py

```diff
@@ -28,6 +28,30 @@ def __init__(self, elemsdict, shape, domain):
         if not all(0 <= c < n for row in self.values() for c in row):
             raise DDMBadInputError("Column out of range")
 
+    def getitem(self, i, j):
+        m, n = self.shape
+        if i >= m:
+                raise IndexError("row index out of range")
+        if j >= n:
+            raise IndexError("column index out of range")
+
+        try:
+            return self[i][j]
+        except KeyError:
+            return self.domain.zero
```

Comment:
I think it's not always necessary to check the bounds on the indices e.g. this might be faster when i and j are indices of a nonzero element:
```
try:
    return self[i][j]
except KeyError:
    m, n = self.shape
    if -m <= i < m and -n <= j < n:
        return self.domain.zero
    else:
        raise IndexError(...)
```

Reply:
Actually that won't handle negative indices so it will need to be changed for that.

Does this work for negative indices? Are there tests for them?

Reply:
@oscarbenjamin, since we need to check for negative indices, then why not to raise in them only. 
This implementation will probably work:
```
def getitem(self, i, j):
        m, n = self.shape
        try:
            if not -m <= i < m:
                raise IndexError("Row index out of range")
            i = i % m
            if not -n <= j < n:
                raise IndexError("Column index out of range")
            j = j % m

            return self[i][j]
        except KeyError:
            return self.domain.zero
```
Kindly share your opinion.
Thanks.

Reply:
That looks like it would work but have you timed the different implementations? Speed is important here.

The cases that need to be fastest are getting a nonzero element and getting a zero element with both positive and negative indices. Raising an error is less important so that case can be slower.

Reply:
Sure, will see the timings.
Thanks.

Reply:
Timing is almost 2x in your implementation:

Your implementation:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DomainMatrix([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [3]: %timeit ddm[9, 9]
713 ns ± 56.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: ddm = DomainMatrix([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [5]: %timeit ddm[9999, 9999]
667 ns ± 68.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
Mine Implementation:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DomainMatrix([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [3]: %timeit ddm[9, 9]
1.02 µs ± 353 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: ddm = DomainMatrix([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [5]: %timeit ddm[9999, 9999]
1.4 µs ± 70.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

This seems to work:
```
def getitem(self, i, j):
        try:
            return self[i][j]
        except KeyError:
            m, n = self.shape
            if -m <= i < m and -n <= j < n:
                try:
                    return self[i % m][j % n]
                except KeyError:
                    return self.domain.zero
            else:
                raise IndexError("index out of range")
```
Timings for this implementation:
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DomainMatrix([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [3]: %timeit ddm[9, 9]
614 ns ± 5.93 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: ddm = DomainMatrix([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [5]: %timeit ddm[9999, 9999]
674 ns ± 43 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

``
Thanks.

Reply:
Looks good but you should also compare timings for negative indices.

Reply:
Timings are almost the same for negative indices as both of them go through the same set of procedures in negative indices:
New implementation: https://github.com/sympy/sympy/pull/21118#discussion_r606241229
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DomainMatrix([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [3]: %timeit ddm[-9999, -9999]
627 ns ± 4.56 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: ddm = DomainMatrix([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [5]: %timeit ddm[-9, -9]
634 ns ± 9.42 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
Old implementation: https://github.com/sympy/sympy/pull/21118#discussion_r606212848
```
In [1]: from sympy.polys.matrices.domainmatrix import DomainMatrix, DomainScalar, DDM,SDM

In [2]: ddm = DomainMatrix([[ZZ(1)]*1000000]*1000000, (1000000, 1000000), ZZ)

In [3]: %timeit ddm[-9999, -9999]
622 ns ± 6.62 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: ddm = DomainMatrix([[ZZ(1)]*10]*10, (10, 10), ZZ)

In [5]: %timeit ddm[-9, -9]
654 ns ± 33.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Should we implement new [method](https://github.com/sympy/sympy/pull/21118#discussion_r606241229)?
Thanks.

Reply:
Yes, looks good